src/goto-instrument/cover.cpp

```diff
@@ -1071,310 +1071,312 @@ void instrument_cover_goals(
                 "<builtin-library-"))
     return;
   
-  const irep_idt coverage_criterion=as_string(criterion);
-  const irep_idt property_class="coverage";
-
-  Forall_goto_program_instructions(i_it, goto_program)
+  for(const auto &criterion:criteria)
   {
-    switch(criterion)
+    const irep_idt coverage_criterion=as_string(criterion);
+    const irep_idt property_class="coverage";
+
+    Forall_goto_program_instructions(i_it, goto_program)
```

Comment:
This approach seems rather inefficient: iterate over all the instructions multiple times? Wouldn't it be possible to turn the criteria into several Booleans and then stick with the previous code that primarily iterates over the instructions? For each instruction, then test, e.g., is_assert() && cover_assertions (with cover_assertions being such a Boolean variable that equals criteria.find(coverage_criteriont::ASSERTION)!=criteria.end().